src/main/java/br/unb/cloudissues/Main.java

```diff
@@ -1,194 +1,185 @@
+
+	// XXX CHANGE THIS LINE. should be your sonar URL until '/api'
+	// example: "https://sonar.eclipse.org/api/"
+	// example: "https://builds.apache.org/analysis/api"
+	// example: "https://sonarcloud.io/api"
+	private static final String SONAR_API_URL = System.getenv("SONAR_API_URL");
+
+	private static final String ISSUES_SEARCH_URL = SONAR_API_URL + "/issues/search";
+
+	private static final String DIRECTORY = "resources/";
+	private static final String FIXED_DIRECTORY = DIRECTORY + "fixed/";
+	private static final String OPEN_DIRECTORY = DIRECTORY + "open-issues/";
+
+	private static final String RULES_LIST = DIRECTORY + "java_rules.json";
+	private static final String PROJECTS_LIST = DIRECTORY + "java_project_list.json";
+
+	private static final String WONT_FIX_FALSE_POSITIVE_DIRECTORY = DIRECTORY
+			+ "wont-fix-false-positive/";
+
+	private static final String FILES_METRICS_DIRECTORY = DIRECTORY + "files-metrics/";
+
+	static {
+		if (SONAR_API_URL == null || SONAR_API_URL.isEmpty())
+			throw new IllegalStateException("Please change the value of SONAR_API_URL");
+	}
+
+	public static void main(String[] args) throws IOException, InterruptedException {
+		retrieveAndWriteRules();
+
+		retrieveAndWriteProjects();
+//		retrieveAndWriteProjectsIfApache();
+
+		requestAndWriteFixedViolationsOneFilePerProject(PROJECTS_LIST, FIXED_DIRECTORY);
+
+		requestAndWriteOpenViolationsOneFilePerProject(PROJECTS_LIST, OPEN_DIRECTORY);
+
+		requestAndWriteFalsePositiveAndWontFixViolations(PROJECTS_LIST,
+				WONT_FIX_FALSE_POSITIVE_DIRECTORY);
+
+		projectsAndFilesMetrics();
+	}
+
+	private static void retrieveAndWriteRules() throws IOException, InterruptedException {
+		JavaRulesRetriever jrr = new JavaRulesRetriever(SONAR_API_URL);
+		List<Rule> rules = jrr.retrieve();
+		Utils.writeObjToFileAsJSON(rules, RULES_LIST);
+	}
+
+	private static void retrieveAndWriteProjects() throws IOException, InterruptedException {
+		JavaProjectsRetriever jpr = new JavaProjectsRetriever.Builder(SONAR_API_URL)
+				.isSonarCloud(false).build();
+		List<Project> sonarJavaProjects = jpr.retrieve();
+		Utils.writeObjToFileAsJSON(sonarJavaProjects, PROJECTS_LIST);
+	}
+
+	static void retrieveAndWriteProjectsIfApache() throws IOException {
+		JavaProjectsCSVRetriever jpr = new JavaProjectsCSVRetriever(SONAR_API_URL + "/projects");
+		Utils.writeObjToFileAsJSON(jpr.retrieve(), PROJECTS_LIST);
+	}
+
+	private static void requestAndWriteFixedViolationsOneFilePerProject(String projectsJsonPath,
+			String fixedDirectory) throws IOException, InterruptedException {
+		System.out.println("\nRetrieving fixed violations");
+
+		List<Project> projects = Utils.retrieveCollectionFromJSONFile(projectsJsonPath,
+				Project.class);
+
+		for (Project project : projects) {
+			requestAndWriteFixedViolationsForProjects(Utils.generateJsonPathToSaveForEachProject(
+					fixedDirectory, project.getProjectName()), Collections.singletonList(project));
+			waitBetweenPartitions(10_000);
```

Comment:
#### Changes Here
 Changing the operator asList to singleList, the motivation is because: 
     the asList is mutable and the operation from methods `requestAndWriteOpenViolationsForProjects` and `requestAndWriteFalsePositiveViolationsForProjects` 
     not handle(inserting, removing, and change, just consuming values)